code/model2.py

```diff
@@ -1,9 +1,129 @@
+import time
 from model import Model
+import torch
+import torch.nn as nn
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
 
 
-class Model2(Model):
+class Model2(nn.Module, Model):
+    """
+    Contains the code for Model3.
+    Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
+     - One sr_embedding is calculated for each email as a concatenation of sender_emb and average of all receiver_
+     emb present in the mail
+     - Loss is calculated as L2 loss between predicted_email_embedding and email_representation. The email
+     representation is obtained by averaging embeddings from pre-trained word2vec model
+    """
+
+    def __init__(self, epochs=10):
+        self.epochs = epochs
+        super(Model2, self).__init__()
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+        # first hidden layer, linear layer with weights 200x500
+        # this should be the size of <sender+receiver representation>
+        self.h1_layer = nn.Linear(200, 500)
+        # ReLU activation used
+        self.relu = nn.ReLU()
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(500, 50)
+
+    def forward(self, s_id, r_ids,pv_emb, nv_emb):
+        """
+        Input is an integer sender_id and a list of receiver_ids. These ids are first converted into torch Variables and
+        then are used to lookup embeddings. If there are multiple receivers, an average embeddings is taken of all the
+        receivers. The gradients in the receiver embeddings are distributed accordingly.
+        Does a forward pass and returns the predicted email representation
+        :param s_id:
+        :param r_id
+        :return: email representation
+        """
+        # convert integers to long tensors
+        s_id = autograd.Variable(torch.LongTensor([s_id]))
+        r_ids = autograd.Variable(torch.LongTensor(r_ids))
+        # extract the embedding for all the receivers and take an average
+        r_emb = torch.mean(self.embedding_layer(r_ids), 0, True)
+        # extract the sender embedding
+        s_emb = self.embedding_layer(s_id)
+        # print s_emb.size()
+        # print r_emb.size()
+        #print size(pv_emb)
+        p_emb = torch.autograd.Variable(torch.from_numpy(pv_emb))
```

Comment:
Although Variable's requires_grad defaults to false, could you just cross-check that p_emb and n_emb are not getting updated during training?